### PR TITLE
lightning: support escaped dump file name

### LINKED ISSUE #31512: tidb-lightning should support escaped scheme/table name
## Feature Request

**Is your feature request related to a problem? Please describe:**

Dumpling has already supported escape the schema/table name if it contains special chars in https://github.com/pingcap/dumpling/pull/140. TiDB Lightning should be compatible with this behavior.

**Describe the feature you'd like:**

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by glorv**
### What problem does this PR solve?

Issue Number: close #31512

Problem Summary:
Support escaped schema/table name in source file for the default file-routing rules. 

NOTE: we only apply an unescape for schema/table name for the default rules, so it can be compatible with dumpling. For customized rules, we leave this responsibility to the user to user the extract names are correct.

### Check List

Tests 

- [x] Unit test
- [] Integration test
  - There is a corresponding integration test in DM
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/config/config.go [hunk 1]
```diff
@@ -1024,6 +1024,10 @@ func (cfg *Config) CheckAndAdjustTiDBPort(ctx context.Context, mustHaveInternalC
 	if cfg.TiDB.Port <= 0 {
 		return errors.New("invalid `tidb.port` setting")
 	}
+	// remove the http(s) schema and we will fill it depend on the tls settings.
+	cfg.TiDB.PdAddr = strings.TrimPrefix(cfg.TiDB.PdAddr, "https://")
+	cfg.TiDB.PdAddr = strings.TrimPrefix(cfg.TiDB.PdAddr, "http://")
```
Comment (gozssky) [inline br/pkg/lightning/config/config.go:1029]: It seems we always use the host part of `PdAddr` or assume it is a host. I think it is more appropriate to use `url.Parse` to parse its host.

File: br/pkg/lightning/mydump/loader.go [hunk 2]
```diff
@@ -290,8 +291,22 @@ func (s *mdLoaderSetup) listFiles(ctx context.Context, store storage.ExternalSto
 			return nil
 		}
 
+		schemaName, err := url.QueryUnescape(res.Schema)
```
Comment (gozssky) [inline br/pkg/lightning/mydump/loader.go:294]: How can we tell if we need to unscape the schema? (`db%22` is also a valid database name)

Comment (glorv) [reply br/pkg/lightning/mydump/loader.go:294]: I think there is not easy to precisely check whether the input file name is escaped. This behavior is similar to our default file routint rule with a file name like `my.db.my.table.001.sql`, we don't know with `.` is the separator between dbname and tablename.

File: br/pkg/lightning/mydump/loader.go [hunk 3]
```diff
@@ -290,8 +291,22 @@ func (s *mdLoaderSetup) listFiles(ctx context.Context, store storage.ExternalSto
...
+		if err != nil {
+			log.L().Warn("decode schema name failed, skip decode and use the origin", zap.String("schema", res.Schema),
```
Comment (kennytm) [inline br/pkg/lightning/mydump/loader.go:296]:
```suggestion
			log.L().Warn("decode schema name failed, skip decode and use the original", zap.String("schema", res.Schema),
```

File: br/pkg/lightning/mydump/loader_test.go [hunk 4]
```diff
@@ -622,3 +622,82 @@ func (s *testMydumpLoaderSuite) TestFileRouting(c *C) {
...
+
+	s.touch(c, "db%22-schema-create.sql")
+	s.touch(c, "db%22.t%2522-schema.sql")
+	s.touch(c, "db%22.t%2522.0.csv")
+
+	mdl, err := md.NewMyDumpLoader(context.Background(), s.cfg)
+	c.Assert(err, IsNil)
+	c.Assert(mdl.GetDatabases(), DeepEquals, []*md.MDDatabaseMeta{
+		{
+			Name:       "db\"",
```
Comment (kennytm) [inline br/pkg/lightning/mydump/loader_test.go:653]:
```suggestion
			Name:       `db"`,
```

File: br/pkg/lightning/mydump/loader_test.go [hunk 5]
```diff
@@ -622,3 +622,82 @@ func (s *testMydumpLoaderSuite) TestFileRouting(c *C) {
...
+
+			db%22.t%2522-schema.sql
+			db%22.t%2522.0.csv
+	*/
+
+	s.touch(c, "test-schema-create.sql")
+	s.touch(c, "test.t%22-schema.sql")
+	s.touch(c, "test.t%22.sql")
+	s.touch(c, "test.t%2522-schema.sql")
+	s.touch(c, "test.t%2522.csv")
```
Comment (kennytm) [inline br/pkg/lightning/mydump/loader_test.go:643]: include the following two cases:

* `test.t%gg.csv`
* `test.t+gg.csv` (with this you should realize that you probably shouldn't use `url.QueryUnescape`)

Comment (glorv) [reply br/pkg/lightning/mydump/loader_test.go:647]: done

Comment (glorv) [reply br/pkg/lightning/mydump/loader_test.go:647]: Replace with `url.PathUnescape`😅

File: br/pkg/lightning/config/config.go (hunk 1 above)
Comment (glorv) [reply br/pkg/lightning/config/config.go:1028]: removed as this is not a complete solution. We may fully resolve this issue in a separate PR.

**Comment by glorv** [comment]
@kennytm @gozssky PTAL again. https://github.com/pingcap/tiflow/pull/4034 depends on this PR to pass the integration tests. 😂

**Review by gozssky — APPROVED** [review]

**Review by kennytm — APPROVED** [review]

**Pull request merged**